libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,225 @@
+using Tags = std::vector<std::pair<std::string, Value>>;
+
+class ProgressScope;
+class ProgressSpan;
+class ProgressContext;
+
+// This class does not currently support thread-local tracers or concurrency controls
+// Functions should only be used in a single-threaded context
+
+// Best Practices:
+//   If possible always avoid creating raw ProgressSpans from ProgressTracer
+//   If possible always use ProgressScopes to handle ProgressSpans
+//   Only use 1 ProgressTracer in an execution, so they should be created at entry points
+//   Use ProgressScope's RAII to handle early returns (i.e. due to errors)
+//   Raw ProgressSpans should be used for special scenarios like tracing asynchronous calls
+
+// Notes:
+//   SharedMemSys and DistMemSys will initialize the global ProgressTracer to a JsonTracer
+//     later it will be set to a No-op Tracer, on Fini or the destructor they will call
+//     Close on the global ProgressTracer
+//   This prevents GetProgressTracer() from returning nullptr and ensures Tracers are closed
+//
+//   ProgressScope's will only close their ProgressSpan when their ProgressSpan is the active span
+//   This means that if a user exclusively uses ProgressScopes,
+//   Then parent ProgressSpans will not be closed until their child ProgressSpan's close
+//   i.e.
+//   {
+//      auto scope1 = tracer->StartActiveSpan("1");
+//      auto scope2 = tracer->StartActiveSpan("2");
+//      if (err) { return; }
+//      scope2.Close();
+//      auto scope3 = tracer->StartActiveSpan("3");
+//   }
+//   Always results in scope2's ProgressSpan finishing before scope1's
+//   and scope3's ProgressSpan finishing before scope1's if there is no error
+
+class KATANA_EXPORT ProgressTracer {
+  friend class ProgressSpan;
+
+  void SetSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
+
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
+  static void SetProgressTracer(std::shared_ptr<ProgressTracer> tracer);
+
+  /// Create a new top level span if one does not already exist.
+  /// Otherwise creates a child span of the current span
+  /// Can specify to not finish the current scope’s span by RAII by
+  /// setting finish_on_close=false
+  ProgressScope StartActiveSpan(
+      const std::string& span_name, bool finish_on_close = true);
+  ProgressScope StartActiveSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of,
+      bool finish_on_close = true);
+
+  /// Set the current active thread-local span to provided span
+  /// Returns a ProgressScope
+  /// Can specify to not finish the span when the scope goes out of scope
+  /// by RAII by setting finish_on_close=false
+  /// If called with a nullptr span,
+  ///    calls StartActiveSpan("error: nullptr span given to SetActiveSpan", finish_on_close)
```

Comment:
Just to prevent having the tracing system throw an error, I can change it to fatal though if you think that's best

Reply:
It would be good to get to a point where nullptr is not a valid way to refer to a span/scope/etc. (if possible). This simplifies the life of users of the API.

Reply:
Alight I'll just make it throw an error then, the only way this could really happen is if a developer deliberately sends a nullptr